dbms/src/Flash/Mpp/MPPTunnel.h

```diff
@@ -101,122 +229,67 @@ class MPPTunnelBase : private boost::noncopyable
     // finish the writing.
     void writeDone();
 
-    std::shared_ptr<mpp::MPPDataPacket> readForLocal();
-
     /// close() finishes the tunnel, if the tunnel is connected already, it will
     /// write the error message to the tunnel, otherwise it just close the tunnel
     void close(const String & reason);
 
     // a MPPConn request has arrived. it will build connection by this tunnel;
-    void connect(Writer * writer_);
+    virtual void connect(PacketWriter * writer);
 
     // wait until all the data has been transferred.
     void waitForFinish();
 
     const ConnectionProfileInfo & getConnectionProfileInfo() const { return connection_profile_info; }
 
-    bool isLocal() const { return is_local; }
+    bool isLocal() const { return mode == TunnelSenderMode::LOCAL; }
+    bool isAsync() const { return mode == TunnelSenderMode::ASYNC_GRPC; }
 
     const LoggerPtr & getLogger() const { return log; }
 
-    // do finish work for consumer, if need_lock is false, it means it has been protected by a mutex lock.
-    void consumerFinish(const String & err_msg, bool need_lock = true);
-
-    bool isSendQueueNextPopNonBlocking() { return send_queue.isNextPopNonBlocking(); }
-
-    // In async mode, do a singe send operation when Writer::TryWrite() succeeds.
-    // In sync mode, as a background task to keep sending until done.
-    void sendJob(bool need_lock = true);
+    TunnelSenderPtr getTunnelSender() { return tunnel_sender; }
+    SyncTunnelSenderPtr getSyncTunnelSender() { return sync_tunnel_sender; }
+    AsyncTunnelSenderPtr getAsyncTunnelSender() { return async_tunnel_sender; }
+    LocalTunnelSenderPtr getLocalTunnelSender() { return local_tunnel_sender; }
 
 private:
-    friend class tests::MPPTunnelTest;
-    friend class tests::TestMPPTunnelBase;
-    // For gtest usage
-    MPPTunnelBase(
-        const String & tunnel_id_,
-        std::chrono::seconds timeout_,
-        int input_steams_num_,
-        bool is_local_,
-        bool is_async_,
-        const String & req_id);
+    friend class tests::TestMPPTunnel;
+    enum class TunnelStatus
+    {
+        Unconnected, // Not connect to any writer, not able to accept new data
+        Connected, // Connected to some writer, accepting data
+        WaitingForSenderFinish, // Accepting all data already, wait for sender to finish
+        Finished // Final state, no more work to do
```

Comment:
How about differentiate `Finished` (finish after connect) and `Cancelled` (close before connect)?

Reply:
Yeah, they are different states, I think they are just not so different, since cancel event may lead to Finished state too:) 
In the other side, it might be more meaningful to differentiate cancel and finished, I'll consider it later.